graceful/graceful.go

```diff
@@ -0,0 +1,86 @@
+package graceful
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+var DefaultShutdownTimeout = time.Second * 60
+
+type GracefulServer struct {
+	server   *http.Server
+	listener net.Listener
+	log      *logrus.Entry
+
+	URL             string
+	ShutdownTimeout time.Duration
+}
+
+func NewGracefulServer(handler http.Handler, log *logrus.Entry) *GracefulServer {
+	return &GracefulServer{
+		server:          &http.Server{Handler: handler},
+		log:             log,
+		listener:        nil,
+		ShutdownTimeout: DefaultShutdownTimeout,
+	}
+}
+
+func (svr *GracefulServer) Bind(addr string) error {
+	l, err := net.Listen("tcp", addr)
+	if err != nil {
+		return err
+	}
+	svr.URL = "http://" + l.Addr().String()
+	svr.listener = l
+	return nil
+}
+
+func (svr *GracefulServer) Listen() error {
+	go svr.listenForShutdownSignal()
+	return svr.server.Serve(svr.listener)
```

Comment:
Remember that this will return as soon as `http.Server.Shutdown()` is called. Should we do anything in this package to make that easier, and not something the user has to remember? I worry about calls to `os.Exit` afterwards. `listenForShutdownSignal` is the method that blocks until the server shutdown is complete or times out.